# Analytic Network Process (ANP) in Multi-Criteria Decision Making

## Introduction

The **Analytic Network Process (ANP)** is an advanced generalization of the Analytic Hierarchy Process (AHP) developed by Thomas L. Saaty. While AHP structures decision problems hierarchically, ANP allows for more complex relationships by modeling them as networks with interdependent elements.

---

## Fundamental Concepts

### 1. Network Structure
Unlike AHP's hierarchical structure, ANP uses a network where:
- **Clusters**: Groups of similar elements (criteria, alternatives, stakeholders)
- **Nodes**: Individual elements within clusters
- **Inner Dependencies**: Relationships within a cluster
- **Outer Dependencies**: Relationships between clusters

### 2. Types of Dependencies

#### Inner Dependencies
- **Hierarchical**: One-way relationships within a cluster
- **Loop**: Feedback relationships within a cluster
- **Network**: Complex interrelationships within a cluster

#### Outer Dependencies
- Relationships between different clusters
- Can be one-way or two-way
- Represent strategic influences

### 3. Supermatrix Formation
The supermatrix captures all relationships in the network:
- Rows and columns represent all elements
- Entries show influence strengths between elements
- Block structure reflects cluster organization

### 4. Priority Derivation
ANP calculates priorities through:
- Local priorities from pairwise comparisons
- Supermatrix formation and weighting
- Limit matrix calculation for global priorities

## How ANP Works

### The Network Structure

ANP extends AHP by allowing for more complex relationships:

1. **Clusters**: Groups of elements (e.g., Benefits, Costs, Opportunities, Risks)
2. **Elements**: Individual criteria or alternatives within clusters
3. **Relationships**: Directed influences between elements

### Mathematical Foundation

#### 1. Pairwise Comparison Matrices
Similar to AHP, ANP uses pairwise comparisons but with additional complexity:

- **Inner Dependence Matrices**: Compare elements within the same cluster
- **Outer Dependence Matrices**: Compare influences between clusters
- **Interdependence Matrices**: Compare elements that influence each other

#### 2. Supermatrix Construction
The supermatrix W is a partitioned matrix structured with the Goal as the first element, followed by criteria and alternatives:

$$W = \begin{bmatrix}
\text{ } & \text{Goal} & \text{Criteria} & \text{Alternatives} \\
Goal & W_{G,G} & W_{G,C} & W_{G,A} \\
Criteria & W_{C,G} & W_{C,C} & W_{C,A} \\
Alternatives & W_{A,G} & W_{A,C} & W_{A,A}
\end{bmatrix} = \begin{bmatrix}
\text{ } & \text{Goal} & \text{Criteria} & \text{Alternatives} \\
\text{Goal} & 0 & 0 & 0 \\
Criteria & W_{C,G} & W_{C,C} & W_{C,A} \\
Alternatives & W_{A,G} & W_{A,C} & W_{A,A}
\end{bmatrix}$$

Where:
- $W_{G,G} = W_{G,C} = W_{G,A} = 0$: **Goal row is all zeros because the Goal is the ultimate objective and doesn't influence other elements**
- $W_{C,G}$: How criteria relate to the goal
- $W_{C,C}$: Interdependencies among criteria
- $W_{C,A}$: How criteria influence alternatives
- $W_{A,G}$: How alternatives relate to the goal
- $W_{A,C}$: How alternatives influence criteria (feedback)
- $W_{A,A}$: Interdependencies among alternatives

**Key Point**: The Goal **column** contains relationships TO the goal (how other elements contribute to achieving it), while the Goal **row** is all zeros since the Goal doesn't influence other elements.

#### 3. Weighted Supermatrix
The weighted supermatrix accounts for cluster priorities:

$$\bar{W} = \begin{bmatrix}
\text{ } & \text{Goal} & \text{Criteria} & \text{Alternatives} \\
\text{Goal} & 0 & 0 & 0 \\
Criteria & a_{CG} \cdot W_{C,G} & a_{CC} \cdot W_{C,C} & a_{CA} \cdot W_{C,A} \\
Alternatives & a_{AG} \cdot W_{A,G} & a_{AC} \cdot W_{A,C} & a_{AA} \cdot W_{A,A}
\end{bmatrix}$$

Where $a_{ij}$ represents the priority weight of cluster j's influence on cluster i.

#### 4. Limit Matrix
The limit matrix gives the final priorities:

$$\lim_{k \to \infty} \bar{W}^k$$

This represents the long-term influence priorities in the network, with the Goal column providing the final priority rankings.

## Steps to Perform ANP

### Step 1: Model the Problem as a Network

1. **Define the Goal**: Clearly state the decision objective (top level)
2. **Identify Clusters**: Group related elements together (Criteria, Alternatives, etc.)
3. **Identify Elements**: Define specific criteria/alternatives within clusters
4. **Determine Relationships**: Identify dependencies and influences
5. **Draw Network Diagram**: Visualize the network structure with Goal at the top

### Step 2: Construct Pairwise Comparison Matrices

#### Goal Level
- The Goal typically doesn't have pairwise comparisons as it's the target
- Focus on how lower-level elements contribute to achieving the Goal

#### Inner Dependence Matrices
For each cluster with inner dependencies:
- Compare elements within the cluster
- Consider how each element influences others in the same cluster
- Create matrices for each controlling element

#### Outer Dependence Matrices
For relationships between clusters:
- Compare the influence of elements from one cluster on another
- Create matrices showing relative importance of influences
- Include how criteria relate to the Goal and how alternatives relate to criteria

### Step 3: Calculate Local Priorities

#### For Each Matrix:
1. **Geometric Mean Method**:
   - Calculate geometric mean of each row
   - Normalize to get priority vector

2. **Eigenvalue Method**:
   - Solve for principal eigenvector
   - Normalize for priorities

### Step 4: Check Consistency

#### Calculate Consistency Measures:
1. **Consistency Index (CI)**: $(\lambda_{max} - n)/(n-1)$
2. **Consistency Ratio (CR)**: $CI/RI$
3. **Acceptable CR**: ≤ 0.10

### Step 5: Construct Supermatrix

1. **Unweighted Supermatrix**: 
   - Place Goal in first row and column (typically zeros except for relationships)
   - Arrange all priority vectors in appropriate blocks
   - Criteria in second block, Alternatives in third block

2. **Weighted Supermatrix**: Weight by cluster priorities

3. **Stochastic Supermatrix**: Column-stochastic normalization

### Step 6: Calculate Limit Matrix

1. **Raise to Power**: Compute $\lim_{k \to \infty} W^k$
2. **Extract Priorities**: Get final global priorities from the Goal column
3. **Synthesize Results**: The Goal column in the limit matrix provides final rankings

### Step 7: Sensitivity Analysis

1. **Weight Sensitivity**: Test impact of cluster weight changes
2. **Relationship Sensitivity**: Test strength of interdependencies
3. **Consistency Sensitivity**: Analyze judgment reliability

### Step 8: Make Decision

1. **Rank Alternatives**: Based on Goal column values in limit matrix
2. **Select Best Option**: Choose highest-ranked alternative from Goal perspective
3. **Document Process**: Record all matrices and assumptions

## ANP Process Flowchart

![ANP Steps](../figures/mermaid/ANP_Steps.png)

# ANP Code Example: Hiring Decision

## Problem Description

We need to hire the best candidate for a sales engineering position. The decision involves:

- **Goal**: Select the best candidate
- **Criteria**: Technical Skills, Sales Skills, Work Experience
- **Alternatives**: Candidate A, Candidate B

### Network Structure
- Criteria may influence each other (interdependencies)
- Candidates may have network effects on criteria evaluation
- The Goal depends on all criteria and alternatives

In [98]:
# Import required libraries
import numpy as np
import pandas as pd
from scipy.linalg import eig

# Display options
pd.set_option("display.float_format", "{:.4f}".format)
np.set_printoptions(precision=4, suppress=True)

In [99]:
# Define the problem structure
goal = "Select Best Candidate"
criteria = ["Technical Skills", "Sales Skills", "Work Experience"]
alternatives = ["Candidate A", "Candidate B"]

# All elements in order: Goal, Criteria, Alternatives
all_elements = [goal] + criteria + alternatives
n_total = len(all_elements)

print(f"Goal: {goal}")
print(f"Criteria: {criteria}")
print(f"Alternatives: {alternatives}")
print(f"Total elements: {n_total}")
print()

Goal: Select Best Candidate
Criteria: ['Technical Skills', 'Sales Skills', 'Work Experience']
Alternatives: ['Candidate A', 'Candidate B']
Total elements: 6



In [100]:
# Helper functions
def calculate_eigenvalue_priorities(matrix):
    """
    Calculate priorities using eigenvalue method
    """
    # Calculate eigenvalues and eigenvectors
    eigenvals, eigenvecs = eig(matrix)

    # Find principal eigenvalue (largest)
    principal_idx = np.argmax(eigenvals.real)
    principal_eigenval = eigenvals[principal_idx].real
    principal_eigenvec = eigenvecs[:, principal_idx].real

    # Normalize eigenvector to get priorities
    priorities = principal_eigenvec / np.sum(principal_eigenvec)

    return priorities, principal_eigenval


def calculate_consistency_ratio(matrix, eigenval):
    """
    Calculate consistency ratio for a comparison matrix
    """
    n = matrix.shape[0]

    # Random Index values (from Saaty's work)
    RI_dict = {
        1: 0.00,
        2: 0.00,
        3: 0.58,
        4: 0.90,
        5: 1.12,
        6: 1.24,
        7: 1.32,
        8: 1.41,
        9: 1.45,
        10: 1.49,
    }

    if n in RI_dict:
        RI = RI_dict[n]
    else:
        RI = 1.49  # For n > 10

    if n <= 2:
        return 0.0, 0.0  # Perfect consistency for n=1,2

    # Consistency Index
    CI = (eigenval - n) / (n - 1)

    # Consistency Ratio
    CR = CI / RI if RI > 0 else 0.0

    return CI, CR

In [101]:
def print_matrix_with_labels(matrix, labels, title):
    """
    Print a matrix with row and column labels
    """
    print(f"\n{title}:")
    print("-" * len(title))
    df = pd.DataFrame(matrix, index=labels, columns=labels)
    print(df.to_string(float_format="%.4f"))
    print()


def print_priorities(priorities, labels, title):
    """
    Print priorities with labels
    """
    print(f"{title}:")
    priorities_dict = {
        label: priority for label, priority in zip(labels, priorities)
    }
    print(f"Priorities: {priorities_dict}")
    return priorities_dict

In [102]:
# Step 1: Criteria Interdependency Matrix

# Matrix: Criteria vs Criteria (3x3)
criteria_interdependency = np.array(
    [
        #   Tech   Sales  Exp
        [1.0, 2.0, 1 / 3],  # Technical Skills
        [1 / 2, 1.0, 1 / 4],  # Sales Skills
        [3.0, 4.0, 1.0],  # Work Experience
    ]
)

print("Step 1: Criteria Interdependency Matrix")
print("Question: Which criterion has more influence on achieving the goal?")
print_matrix_with_labels(
    criteria_interdependency, criteria, "Criteria Interdependency Matrix"
)

# Calculate priorities and check consistency
priorities_criteria, max_eig_criteria = calculate_eigenvalue_priorities(
    criteria_interdependency
)
ci_criteria, cr_criteria = calculate_consistency_ratio(
    criteria_interdependency, max_eig_criteria
)

print_priorities(priorities_criteria, criteria, "Criteria Priorities")
print(f"CR = {cr_criteria:.4f}")
print(
    f"Consistency: {'Acceptable' if cr_criteria <= 0.10 else 'Needs revision'}"
)
print("=" * 50)

Step 1: Criteria Interdependency Matrix
Question: Which criterion has more influence on achieving the goal?

Criteria Interdependency Matrix:
-------------------------------
                  Technical Skills  Sales Skills  Work Experience
Technical Skills            1.0000        2.0000           0.3333
Sales Skills                0.5000        1.0000           0.2500
Work Experience             3.0000        4.0000           1.0000

Criteria Priorities:
Priorities: {'Technical Skills': np.float64(0.23848712266284522), 'Sales Skills': np.float64(0.1364998029888613), 'Work Experience': np.float64(0.6250130743482935)}
CR = 0.0158
Consistency: Acceptable


In [103]:
# Step 2: Alternatives Comparison Matrices

# Alternatives comparison with respect to Technical Skills
alt_technical = np.array(
    [
        #    A     B
        [1.0, 3.0],  # Candidate A
        [1 / 3, 1.0],  # Candidate B
    ]
)

# Alternatives comparison with respect to Sales Skills
alt_sales = np.array(
    [
        #    A     B
        [1.0, 1 / 2],  # Candidate A
        [2.0, 1.0],  # Candidate B
    ]
)

# Alternatives comparison with respect to Work Experience
alt_experience = np.array(
    [
        #    A     B
        [1.0, 2.0],  # Candidate A
        [1 / 2, 1.0],  # Candidate B
    ]
)

# Calculate priorities for each comparison
print("Step 2: Alternatives Comparison Matrices")
print("=" * 45)

# Technical Skills
print_matrix_with_labels(
    alt_technical, alternatives, "Alternatives vs Technical Skills"
)
priorities_tech, max_eig_tech = calculate_eigenvalue_priorities(alt_technical)
ci_tech, cr_tech = calculate_consistency_ratio(alt_technical, max_eig_tech)
print_priorities(priorities_tech, alternatives, "Technical Skills Priorities")
print(f"CR = {cr_tech:.4f}")
print()

# Sales Skills
print_matrix_with_labels(
    alt_sales, alternatives, "Alternatives vs Sales Skills"
)
priorities_sales, max_eig_sales = calculate_eigenvalue_priorities(alt_sales)
ci_sales, cr_sales = calculate_consistency_ratio(alt_sales, max_eig_sales)
print_priorities(priorities_sales, alternatives, "Sales Skills Priorities")
print(f"CR = {cr_sales:.4f}")
print()

# Work Experience
print_matrix_with_labels(
    alt_experience, alternatives, "Alternatives vs Work Experience"
)
priorities_exp, max_eig_exp = calculate_eigenvalue_priorities(alt_experience)
ci_exp, cr_exp = calculate_consistency_ratio(alt_experience, max_eig_exp)
print_priorities(priorities_exp, alternatives, "Work Experience Priorities")
print(f"CR = {cr_exp:.4f}")
print("=" * 50)

Step 2: Alternatives Comparison Matrices

Alternatives vs Technical Skills:
--------------------------------
             Candidate A  Candidate B
Candidate A       1.0000       3.0000
Candidate B       0.3333       1.0000

Technical Skills Priorities:
Priorities: {'Candidate A': np.float64(0.75), 'Candidate B': np.float64(0.25)}
CR = 0.0000


Alternatives vs Sales Skills:
----------------------------
             Candidate A  Candidate B
Candidate A       1.0000       0.5000
Candidate B       2.0000       1.0000

Sales Skills Priorities:
Priorities: {'Candidate A': np.float64(0.3333333333333334), 'Candidate B': np.float64(0.6666666666666666)}
CR = 0.0000


Alternatives vs Work Experience:
-------------------------------
             Candidate A  Candidate B
Candidate A       1.0000       2.0000
Candidate B       0.5000       1.0000

Work Experience Priorities:
Priorities: {'Candidate A': np.float64(0.6666666666666667), 'Candidate B': np.float64(0.3333333333333333)}
CR = 0.0000


In [104]:
# Step 3: Alternatives Interdependency

# How alternatives influence each other (networking effects)
alt_interdependency = np.array(
    [
        #    A     B
        [1.0, 1 / 2],  # Candidate A
        [2.0, 1.0],  # Candidate B
    ]
)

print("Step 3: Alternatives Interdependency Matrix")
print(
    "Question: Which candidate has more influence on the other (networking, team dynamics)?"
)
print_matrix_with_labels(
    alt_interdependency, alternatives, "Alternatives Interdependency"
)

priorities_alt_inter, max_eig_alt = calculate_eigenvalue_priorities(
    alt_interdependency
)
ci_alt, cr_alt = calculate_consistency_ratio(alt_interdependency, max_eig_alt)

print_priorities(
    priorities_alt_inter,
    alternatives,
    "Alternatives Interdependency Priorities",
)
print(f"CR = {cr_alt:.4f}")
print("=" * 50)

Step 3: Alternatives Interdependency Matrix
Question: Which candidate has more influence on the other (networking, team dynamics)?

Alternatives Interdependency:
----------------------------
             Candidate A  Candidate B
Candidate A       1.0000       0.5000
Candidate B       2.0000       1.0000

Alternatives Interdependency Priorities:
Priorities: {'Candidate A': np.float64(0.3333333333333334), 'Candidate B': np.float64(0.6666666666666666)}
CR = 0.0000


In [105]:
# Step 4: Cluster Comparison

# Define clusters: Goal, Criteria, Alternatives
cluster_labels = ["Goal", "Criteria", "Alternatives"]

# Cluster comparison matrix - which cluster is more important for the network
# In ANP, we typically focus on how criteria and alternatives interact
cluster_comparison = np.array(
    [
        #   Goal  Criteria  Alternatives
        [0.0, 0.0, 0.0],  # Goal - doesn't influence others
        [1.0, 1.0, 3.0],  # Criteria - more important than alternatives
        [0.1, 1 / 3, 1.0],  # Alternatives - less important than criteria
    ]
)

print("Step 4: Cluster Comparison Matrix")
print("Question: Which cluster has more influence in the network?")
print_matrix_with_labels(
    cluster_comparison, cluster_labels, "Cluster Comparison Matrix"
)

# Calculate cluster priorities (normalized by column)
column_sums = np.sum(cluster_comparison, axis=0)
cluster_weights = {}
for i, cluster in enumerate(cluster_labels):
    if column_sums[i] > 0:
        cluster_weights[cluster] = cluster_comparison[:, i] / column_sums[i]
    else:
        cluster_weights[cluster] = np.zeros(3)

print("Cluster Weights (normalized by column):")
for cluster, weights in cluster_weights.items():
    print(f"{cluster}: {weights}")

print("=" * 50)

Step 4: Cluster Comparison Matrix
Question: Which cluster has more influence in the network?

Cluster Comparison Matrix:
-------------------------
               Goal  Criteria  Alternatives
Goal         0.0000    0.0000        0.0000
Criteria     1.0000    1.0000        3.0000
Alternatives 0.1000    0.3333        1.0000

Cluster Weights (normalized by column):
Goal: [0.     0.9091 0.0909]
Criteria: [0.   0.75 0.25]
Alternatives: [0.   0.75 0.25]


In [106]:
# Step 5: Construct Unweighted Supermatrix

# Initialize 6x6 supermatrix (Goal + 3 Criteria + 2 Alternatives)
supermatrix = np.zeros((n_total, n_total))

# Goal row remains zeros (Goal doesn't influence anything - it's the target)
# Goal column will be filled with relationships TO the goal

# Criteria relationships (rows 1-3, columns 1-3)
# How criteria influence each other
supermatrix[1:4, 1:4] = criteria_interdependency

# Alternatives to Criteria relationships (rows 4-5, columns 1-3)
# How alternatives relate to criteria (using calculated priorities)
supermatrix[4:6, 1] = priorities_tech  # Technical Skills
supermatrix[4:6, 2] = priorities_sales  # Sales Skills
supermatrix[4:6, 3] = priorities_exp  # Work Experience

# Alternatives interdependencies (rows 4-5, columns 4-5)
supermatrix[4:6, 4:6] = alt_interdependency

# Criteria to Goal relationships (rows 1-3, column 0)
# How criteria contribute to the goal
supermatrix[1:4, 0] = priorities_criteria

# Alternatives to Goal relationships (rows 4-5, column 0)
# Direct contribution of alternatives to goal (often small or zero in ANP)
supermatrix[4:6, 0] = [0.1, 0.1]  # Small direct contribution

# Ensure Goal row is all zeros (Goal doesn't influence other elements)
supermatrix[0, :] = 0

print("Step 5: Unweighted Supermatrix")
print("=" * 40)

# Display supermatrix with labels
supermatrix_df = pd.DataFrame(
    supermatrix, index=all_elements, columns=all_elements
)
supermatrix_df

Step 5: Unweighted Supermatrix


Select Best Candidate  Technical Skills  Sales Skills  \
Select Best Candidate                 0.0000            0.0000        0.0000   
Technical Skills                      0.2385            1.0000        2.0000   
Sales Skills                          0.1365            0.5000        1.0000   
Work Experience                       0.6250            3.0000        4.0000   
Candidate A                           0.1000            0.7500        0.3333   
Candidate B                           0.1000            0.2500        0.6667   

                       Work Experience  Candidate A  Candidate B  
Select Best Candidate           0.0000       0.0000       0.0000  
Technical Skills                0.3333       0.0000       0.0000  
Sales Skills                    0.2500       0.0000       0.0000  
Work Experience                 1.0000       0.0000       0.0000  
Candidate A                     0.6667       1.0000       0.5000  
Candidate B                     0.3333       2.0000       1.0000

In [107]:
# Step 6: Create Weighted Supermatrix

# Extract calculated cluster weights from Step 4
criteria_cluster_weight = cluster_weights["Criteria"][
    1
]  # How criteria influence criteria block
alternatives_cluster_weight = cluster_weights["Alternatives"][
    2
]  # How alternatives influence alternatives block

# Create weighted supermatrix using calculated weights
weighted_supermatrix = supermatrix.copy()

# Apply cluster weights to respective blocks
# Column 0 (Goal): No weighting needed - it's the target
# Columns 1-3 (Criteria): Weight by criteria cluster importance
# Columns 4-5 (Alternatives): Weight by alternatives cluster importance
weighted_supermatrix[:, 1:4] *= criteria_cluster_weight
weighted_supermatrix[:, 4:6] *= alternatives_cluster_weight

# Normalize columns to be s
# tochastic (column sums = 1)
for j in range(n_total):
    col_sum = np.sum(weighted_supermatrix[:, j])
    if col_sum > 0:
        weighted_supermatrix[:, j] = weighted_supermatrix[:, j] / col_sum

print("Step 6: Weighted Supermatrix")
print("=" * 35)

# Display weighted supermatrix
weighted_supermatrix_df = pd.DataFrame(
    weighted_supermatrix, index=all_elements, columns=all_elements
)
weighted_supermatrix_df

Step 6: Weighted Supermatrix


Select Best Candidate  Technical Skills  Sales Skills  \
Select Best Candidate                 0.0000            0.0000        0.0000   
Technical Skills                      0.1987            0.1818        0.2500   
Sales Skills                          0.1137            0.0909        0.1250   
Work Experience                       0.5208            0.5455        0.5000   
Candidate A                           0.0833            0.1364        0.0417   
Candidate B                           0.0833            0.0455        0.0833   

                       Work Experience  Candidate A  Candidate B  
Select Best Candidate           0.0000       0.0000       0.0000  
Technical Skills                0.1290       0.0000       0.0000  
Sales Skills                    0.0968       0.0000       0.0000  
Work Experience                 0.3871       0.0000       0.0000  
Candidate A                     0.2581       0.3333       0.3333  
Candidate B                     0.1290       0.6667       0.6667

In [108]:
# Step 7: Calculate Limit Matrix

# The limit matrix is computed by raising the weighted supermatrix to a large power
# This converges to the long-term influence priorities


def calculate_limit_matrix(matrix, max_iterations=100, tolerance=1e-6):
    """
    Calculate limit matrix by repeated multiplication
    """
    current_matrix = matrix.copy()

    for i in range(max_iterations):
        next_matrix = np.linalg.matrix_power(current_matrix, 2)

        # Check convergence
        if np.allclose(current_matrix, next_matrix, atol=tolerance):
            print(f"Converged after {i + 1} iterations")
            break

        current_matrix = next_matrix

    return current_matrix


print("Step 7: Calculate Limit Matrix")

# Calculate limit matrix
limit_matrix = calculate_limit_matrix(weighted_supermatrix)

# Display limit matrix with labels
limit_matrix_df = pd.DataFrame(
    limit_matrix, index=all_elements, columns=all_elements
)
print("\nLimit Matrix:")
print(limit_matrix_df.to_string(float_format="%.4f"))

# Extract final priorities from Goal column (column 0)
final_priorities = limit_matrix[:, 0]

print("\nFinal Priorities (from Goal column):")
for i, element in enumerate(all_elements):
    print(f"{element:<20}: {final_priorities[i]:.4f}")

Step 7: Calculate Limit Matrix
Converged after 7 iterations

Limit Matrix:
                       Select Best Candidate  Technical Skills  Sales Skills  Work Experience  Candidate A  Candidate B
Select Best Candidate                 0.0000            0.0000        0.0000           0.0000       0.0000       0.0000
Technical Skills                      0.0000            0.0000        0.0000           0.0000       0.0000       0.0000
Sales Skills                          0.0000            0.0000        0.0000           0.0000       0.0000       0.0000
Work Experience                       0.0000            0.0000        0.0000           0.0000       0.0000       0.0000
Candidate A                           0.3333            0.3333        0.3333           0.3333       0.3333       0.3333
Candidate B                           0.6667            0.6667        0.6667           0.6667       0.6667       0.6667

Final Priorities (from Goal column):
Select Best Candidate: 0.0000
Technical Skills 

In [109]:
# Step 8: Final Results and Analysis

print("Step 8: Final Results and Decision Analysis")

# Extract only alternatives priorities
alternatives_priorities = final_priorities[
    4:6
]  # Last 2 elements are alternatives

# Create results summary
results_df = pd.DataFrame(
    {
        "Alternative": alternatives,
        "Priority": alternatives_priorities,
        "Percentage": alternatives_priorities * 100,
        "Rank": [
            1
            if alternatives_priorities[0] > alternatives_priorities[1]
            else 2,
            2
            if alternatives_priorities[0] > alternatives_priorities[1]
            else 1,
        ],
    }
)

print("\nFINAL RANKING:")
results_df_sorted = results_df.sort_values("Priority", ascending=False)
print(results_df_sorted.to_string(index=False, float_format="%.4f"))

# Best alternative
best_alternative = results_df_sorted.iloc[0]["Alternative"]
best_score = results_df_sorted.iloc[0]["Priority"]

print("\nRECOMMENDATION:")
print(f"Best candidate: {best_alternative}")
print(f"Priority score: {best_score:.4f} ({best_score * 100:.1f}%)")

# Show individual criteria performance
print("\nDETAILED ANALYSIS:")

print("\nCriteria Priorities (from Step 1):")
for i, criterion in enumerate(criteria):
    print(f"  {criterion:<18}: {priorities_criteria[i]:.4f}")

print("\nAlternative Performance by Criteria:")
print(f"{'Criterion':<18} {'Candidate A':<12} {'Candidate B':<12}")
print(
    f"{'Technical Skills':<18} {priorities_tech[0]:<12.4f} {priorities_tech[1]:<12.4f}"
)
print(
    f"{'Sales Skills':<18} {priorities_sales[0]:<12.4f} {priorities_sales[1]:<12.4f}"
)
print(
    f"{'Work Experience':<18} {priorities_exp[0]:<12.4f} {priorities_exp[1]:<12.4f}"
)

print("\nNetwork Effects (Interdependencies):")
print(f"{'Alternative':<15} {'Influence Score':<15}")
print(f"{'Candidate A':<15} {priorities_alt_inter[0]:<15.4f}")
print(f"{'Candidate B':<15} {priorities_alt_inter[1]:<15.4f}")

print("\nKey Insights:")
print(
    f"1. Work Experience is the most important criterion ({priorities_criteria[2]:.1%})"
)
print(
    f"2. Candidate B has stronger network influence ({priorities_alt_inter[1]:.1%})"
)
print("3. ANP considers both direct performance and network effects")
print(
    "4. Final decision incorporates interdependencies between criteria and alternatives"
)

Step 8: Final Results and Decision Analysis

FINAL RANKING:
Alternative  Priority  Percentage  Rank
Candidate B    0.6667     66.6667     1
Candidate A    0.3333     33.3333     2

RECOMMENDATION:
Best candidate: Candidate B
Priority score: 0.6667 (66.7%)

DETAILED ANALYSIS:

Criteria Priorities (from Step 1):
  Technical Skills  : 0.2385
  Sales Skills      : 0.1365
  Work Experience   : 0.6250

Alternative Performance by Criteria:
Criterion          Candidate A  Candidate B 
Technical Skills   0.7500       0.2500      
Sales Skills       0.3333       0.6667      
Work Experience    0.6667       0.3333      

Network Effects (Interdependencies):
Alternative     Influence Score
Candidate A     0.3333         
Candidate B     0.6667         

Key Insights:
1. Work Experience is the most important criterion (62.5%)
2. Candidate B has stronger network influence (66.7%)
3. ANP considers both direct performance and network effects
4. Final decision incorporates interdependencies between cri

## ANP vs AHP Comparison

The Analytic Network Process (ANP) extends AHP by considering interdependencies and feedback loops in the decision network. Here's how the results compare:

### Key Differences

1. **Structure**:
   - **AHP**: Hierarchical structure with one-way relationships
   - **ANP**: Network structure with interdependencies and feedback

2. **Relationships**:
   - **AHP**: Criteria are independent; alternatives only relate to criteria
   - **ANP**: Criteria can influence each other; alternatives can have network effects

3. **Complexity**:
   - **AHP**: Simpler calculation with direct priority synthesis
   - **ANP**: More complex with supermatrix and limit matrix calculations

### Results Interpretation

In our hiring example:

- **Candidate B wins** with 66.7% priority due to:
  - Strong performance in Sales Skills (66.7% vs 33.3%)
  - Higher network influence (team dynamics effect)
  - Balanced performance across criteria despite weaker technical skills

- **Work Experience dominates** with 62.5% importance, showing that experience is the most critical factor for this sales engineering position

- **Network effects matter**: Candidate B's stronger interpersonal influence (66.7%) contributes to the final decision

### When to Use ANP vs AHP

**Use ANP when**:
- Criteria have interdependencies
- Alternatives influence each other (network effects)
- Feedback loops exist in the decision system
- Complex organizational or strategic decisions

**Use AHP when**:
- Criteria are independent
- Clear hierarchical structure exists
- Simpler decision problems
- Quick analysis needed